gemini/mcp/mcp_orchestration_app/src/gemini_client.py

```diff
@@ -0,0 +1,781 @@
+        """Sets the generation configuration for subsequent calls."""
+        self._generation_config = config
+        logging.debug(f"Generation config set: {config}")
+
+    def set_system_instruction(self, system_instruction: str) -> None:
+        """Sets the system instruction and initializes the chat session."""
+        self._initialize_client()  # Ensure client is ready
+        if (
+            not self._client
+        ):  # Should not happen if _initialize_client works, but belt-and-suspenders
+            raise ConnectionError("LLM Client not initialized.")
+
+        try:
+            # Create a new chat session with the system instruction
+            # Note: Depending on the exact gen ai version, system instructions might be
+            # handled differently (e.g., as the first message). This assumes `create`
+            # handles it or we send it as the first message implicitly.
+            # Let's assume the original pattern of sending it first was intended.
+            self._chat_session = self._client.chats.create(model=self.model_name)
+            self._chat_session.send_message(system_instruction)
+            # Sending the system instruction as the first message
+            # Important: Check if the library expects a specific format or role for system instructions
+            # If `create` supports `system_instruction=` kwarg, use that instead.
+            # For now, mimicking the original code's behavior.
+            # A more robust approach would use the library's designated system instruction mechanism if available.
+            # self._chat_session.send_message(system_instruction) # Sending it here might consume the first turn
+            logging.info(
+                "LLM chat session initialized."
+            )  # System instruction set is implied
+
+        except Exception as e:
+            logging.exception(
+                "Failed to create LLM chat session or set system instruction."
+            )
+            self._chat_session = None  # Ensure session is None on failure
+            raise ConnectionError("Could not create LLM chat session.") from e
+
+    @staticmethod
+    def extract_tool_call_json(text: str) -> Optional[Dict[str, Any]]:
+        """
+        Extracts a JSON object formatted for tool calls from markdown code blocks.
+
+        Specifically looks for ```json { "tool": ..., "arguments": ... } ```
+
+        Args:
+            text: The input string potentially containing the JSON tool call.
+
+        Returns:
+            The loaded Python dictionary representing the tool call, or None if
+            extraction/parsing fails or if it's not a valid tool call structure.
+        """
+        # Regex to find ```json ... ``` block
+        # Using non-greedy matching .*? for the content
+        match = re.search(r"```json\s*(\{.*?\})\s*```", text, re.DOTALL | re.IGNORECASE)
+        json_string = None
+
```

Comment:
You shouldn't need to do this regex searching if you set `GenerateContentConfig` with `response_mime_type` of `application/json` or if you use Function Calling